presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.spi.Page;
+
+import java.io.Closeable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface Spiller
+        extends Closeable
+{
+    /**
+     * Initiate spilling of pages stream. Returns completed future once spilling has finished.
+     */
+    CompletableFuture<?> spill(Iterator<Page> pageIterator);
```

Comment:
Let's not overgeneralize it at this point. We can refactor it later, when such thing will be indeed required.

For joins it seems like we need different implementation of spillers. You can look at our WiP interfaces for join here:
https://github.com/Teradata/presto/pull/356/commits/db6a9b7ca9fa231a7c923fc7af20772c61366afb
and here:
https://github.com/Teradata/presto/pull/356/commits/97c3ebe1baa4f03a6139c877268a9a4e1c687820

We are thinking about later reimplementing this spiller (used in aggregations) so that it will use `SingleStreamSpiller` interface (similarly as `GenericPartitioningSpiller` can wrap any given `SingleStreamSpiller` to provide `PartitioningSpiller`).

I don't see an advantage of having `Slice` as an interface. It would only complicate things for both joins and aggregations.